ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/StratumConnection.java

```diff
@@ -60,23 +64,49 @@ void handleBuffer(final Buffer buffer) {
       closeHandle.run();
       return;
     }
-    Iterator<String> messages = splitter.split(messagesString).iterator();
-    while (messages.hasNext()) {
-      String message = messages.next();
-      if (!firstMessage) {
-        message = incompleteMessage + message;
-        firstMessage = true;
+
+    if (httpDetected) {
+      httpDetected = false;
+      messagesString = incompleteMessage + messagesString;
+    }
+    Matcher match = contentLengthPattern.matcher(messagesString);
+    if (match.find()) {
+      try {
+        int contentLength = Integer.parseInt(match.group(1));
+        String body = messagesString.substring(messagesString.indexOf("\r\n\r\n") + 4);
+        if (body.length() < contentLength) {
+          incompleteMessage = messagesString;
+          httpDetected = true;
+          return;
+        }
+      } catch (NumberFormatException e) {
+        close();
+        return;
       }
-      if (!messages.hasNext()) {
-        incompleteMessage = message;
-      } else {
-        LOG.trace("Dispatching message {}", message);
-        handleMessage(message);
+
+      LOG.trace("Dispatching HTTP message {}", messagesString);
+      handleMessage(messagesString);
+    } else {
+      boolean firstMessage = false;
+      Splitter splitter = Splitter.on('\n');
+      Iterator<String> messages = splitter.split(messagesString).iterator();
+      while (messages.hasNext()) {
+        String message = messages.next();
+        if (!firstMessage) {
+          message = incompleteMessage + message;
+          firstMessage = true;
+        }
+        if (!messages.hasNext()) {
+          incompleteMessage = message;
+        } else {
+          LOG.trace("Dispatching message {}", message);
+          handleMessage(message);
+        }
       }
     }
   }
 
-  void close(final Void aVoid) {
+  void close() {
     if (protocol != null) {
       protocol.onClose(this);
     }
```

Comment:
Possibly out of scope, but in `handleMessage()` (next method) if the protocol is null the protocol is set but then the message is not handled.  `canHandle` is called but handling is never performed.  should `canHandle` be `maybeHandle` if it also conditionally handles?

Reply:
That's a better name, yes. It changed this time around since this new protocol needs to handle the first message, when others would simply ignore it.